File: vehicle-tracker-backend/vehicle-service/src/main/java/com/github/diogocerqueiralima/infrastructure/mappers/VehicleMapper.java
package com.github.diogocerqueiralima.infrastructure.mappers;

import com.github.diogocerqueiralima.domain.model.Vehicle;
import com.github.diogocerqueiralima.infrastructure.entities.VehicleEntity;
import org.springframework.stereotype.Component;

@Component("infrastructure")
public class VehicleMapper {

    public VehicleEntity toEntity(Vehicle vehicle) {

        VehicleEntity vehicleEntity = new VehicleEntity();

        vehicleEntity.setId(vehicle.getId());
        vehicleEntity.setVin(vehicle.getVin());
        vehicleEntity.setPlate(vehicle.getPlate());
        vehicleEntity.setModel(vehicle.getModel());
        vehicleEntity.setManufacturer(vehicle.getManufacturer());
        vehicleEntity.setYear(vehicle.getYear());
        vehicleEntity.setOwnerId(vehicle.getOwnerId());

        return vehicleEntity;
    }

    public Vehicle toDomain(VehicleEntity vehicleEntity) {
        return new Vehicle(
                vehicleEntity.getId(),
                vehicleEntity.getVin(),
                vehicleEntity.getPlate(),
                vehicleEntity.getModel(),
                vehicleEntity.getManufacturer(),
                vehicleEntity.getYear(),
                vehicleEntity.getOwnerId()
        );
    }

}

File: vehicle-tracker-mobile/app/src/main/java/com/github/diogocerqueiralima/infrastructure/mappers/VehicleMapper.kt
package com.github.diogocerqueiralima.infrastructure.mappers

import com.github.diogocerqueiralima.domain.model.Vehicle
import com.github.diogocerqueiralima.infrastructure.entities.VehicleEntity
import java.time.LocalDate

/**
 * Extension function to convert a [VehicleEntity] infrastructure entity to a [Vehicle] domain model.
 */
fun VehicleEntity.toDomain() = Vehicle(
    vin = this.vin,
    plate = this.plate,
    model = this.model,
    manufacturer = this.manufacturer,
    manufacturingDate = LocalDate.parse(this.manufacturingDate)
)